### ISSUE TITLE
Separate Compilation

### ISSUE THREAD

**Issue by Vertmo** [carp-lang/Carp#889]
##### Disclaimer : maybe I've missed something and this is already in the work, but I haven't been able to find anything in the issues / PRs. Sorry if this issue is stupid/redundant

As I explore and write more with the language, one drawback of the current compiler is apparent : it does not support separate compilation and linking, which mainly means that any used library are compiled from source at the same time as the program. This makes the compilation of any program slower (both for the *carp* compiler and the *c* compiler which has to handle a big C file), and I imagine it will get slower as the size of the standard library increases.

A solution to this problem is to allow for separate compilation : that is allowing to compile a unit of code (let's say a file) to an intermediary representation that can be linked to the complete program. This will allow to compile the core library (or any other) once, and being able to link it when using it into other programs.

Here is my precise proposal for carp: when compiling a project, add an option `--as-lib` to compile as a linkable library. The compiler would then generate three artifacts:
* a *.o* file containing the implementation of the library (compiled as normal, the only difference beeing asking the c compiler to not link it)
* a *.h* interface file that will be used by the C compiler when treating a program using the library
* a *.carpi* (?) interface file that will be used by the Carp compiler when treating a program using the library. This last one will mainly contain the types of the values defined in the source file. I'm not yet sure what format this one would have (text or binary). The second one would probably be more efficient

These artifacts could then be used when loading a compiled library, with a directive such as `(static-load "lib.carpi")` which would tell the compiler to:
* use the *lib.carpi* interface when type/borrow-checking the current program
*  `#include "lib.h"` in the generated c file
* ask the C compiler to link *lib.o* when linking the final executable

Now there may be something I'm missing (in particular regarding the interactions between the repl and a statically compiled library), but this seems doable (although it would need a bit of work). Please tell me what you think ^^

**Comment by hellerve** [comment]
I think this is a great idea, and something that makes a lot of sense. I would definitely want to keep completely static compilation as an option, simply becausde you don’t always have `.o` files etc. lying around, especially for cross-compilation, and having a single artifact is a little more convenient then.

But I see no problem making the two co-exist—it’s just a bunch of work to add that, and I think that’s why we stayed away from it until now. The current approach, while suboptimal for compile times, “works”.

I’d love to investigate all kinds of avenues to make the compiler faster, though, and separate compilation seems like a good way to get faster compile times, easily.

**Comment by jacereda** [comment]
I think it would be better to optimise the haskell side first. Slow C compiles can be solved quite well by using another compilers. TCC is like 10 times faster than clang and there're other interesting projects like https://git.sr.ht/~mcf/cproc (haven't tried it yet).

I'm using tcc and all I can say is that compile times for the C side are blazingly fast on a 11 years old box.

**Comment by Vertmo** [comment]
@jacereda you're kind of right, that's why I would begin with implementing separate compilation at the Carp compiler level, since the cost of running the C compiler is not that high. Although once that's done it would probably not be that difficult to also avoid repeated run of the C compiler.

Well, in any case I will start looking into it. I'll update you if/when I make some progress

**Comment by hellerve** [comment]
Right, I don’t think the C compiler part needs to be optimized so much. What would be way more interesting would incremental/separate compilation of Carp libs/files! That’s how I understood the issue anyway.

**Comment by Vertmo** [comment]
Ok. I've given it a bit more thought (and started writing a bit of code) and I do see an issue in this approach : some things can't just be compiled away to C and then reused at the Carp level. To be exact, these things are:
- macros (and dynamic functions) which are executed at compile time
- polymorphic functions, which are only compiled when used, with their type parameters instantiated (which makes sense as C doesn't support polymorphism)
I'm still not sure how to threat these. It would be stupid to just "drop" them when using separate compilation, as these are both major features of the language. 
However, I will start working on this project ignoring this issue (dropping them then), and comeback to it later with ideas (simply because it's always better to start with the simple cases). If anybody has an idea about it, I'm interested (the first solution I see is to keep them intact in the *.carpi* file so they can be compiled when used).

**Comment by eriksvedang** [comment]
Yeah, it's a tricky problem. Looking forward to seeing what you come up with!
